asterius/src/Asterius/Internals/Parallel.hs

```diff
@@ -0,0 +1,47 @@
+module Asterius.Internals.Parallel
+  ( parallelFor,
+    parallelFor_,
+  )
+where
+
+import Control.Concurrent
+import Control.Concurrent.MVar
+import Control.Monad
+
+-- | Given the worker thread pool capacity @c@, @parallelFor c xs f@ maps @f@
+-- on @xs@ in parallel on the global thread pool (the results are ignored). If
+-- @c = 1@ then it is equivalent to @mconcat <$> mapM fn xs@, thus avoiding
+-- threading overhead.
+{-# INLINE parallelFor_ #-}
+parallelFor_ :: Int -> [a] -> (a -> IO r) -> IO ()
+parallelFor_ n xs fn
+  | n >= 2 = error "TODO"
+  -- If there are not enough resources, fall back to the sequential version.
+  | otherwise = mapM_ fn xs
+
+-- | Given the worker thread pool capacity @c@, @parallelFor c xs f@ maps @f@
+-- on @xs@ in parallel on the global thread pool, and concatenates the results.
+-- If @c = 1@ then it is equivalent to @mconcat <$> mapM fn xs@, thus avoiding
+-- threading overhead.
+{-# INLINE parallelFor #-}
+parallelFor :: Monoid r => Int -> [a] -> (a -> IO r) -> IO r
```

Comment:
> This splits the input to different chunks and distribute them to the worker threads. IMO this has higher risk of unevenly distribute work.

That's true, with this approach we get idle threads sometimes.

> A better approach is using an `IORef [a]` to coordinate between the worker threads, each thread using `atomicModifyIORef'` to dynamically fetch an element and do the computation (and exit once there's no more elements).

Indeed, this wastes no resources, but it also is really concurrent, possibly changing the order of things. `Monoid r` gives us associativity but what you're suggesting also needs symmetry to be correct. Is it the case that `m1 <> m2 == m2 <> m1` for `AsteriusModule`s?

Reply:
We don't need to worry about ordering of `<>`; we never concat two `AsteriusModule`s which contain the same symbols.

Reply:
Right. I point out the symmetry requirement in the function description anyway (since the function is parametric over `r`), to be fair. So, does bd7bfa6b15c1dc70b764856e8638a788f2062149 seem to implement what you have in mind?

Reply:
See latest comments.